File: cpp/567.cpp
class Solution {
private:
    bool permutation(int* x, int* y) {
        for (int i = 0; i < 26; i++)
            if (x[i] != y[i])
                return false;
        return true;
    }

public:
    bool checkInclusion(string& x, string& y) {
        int l = x.length(), m = y.length();
        if (l > m)
            return false;
        int i[26] = {0};
        int j[26] = {0};
        for (int k = 0; k < l; k++) {
            i[x[k] - 97]++;
            j[y[k] - 97]++;
        }
        if (permutation(i, j))
            return true;
        for (int k = l; k < m; k++) {
            j[y[k - l] - 97]--;
            j[y[k] - 97]++;
            if (permutation(i, j))
                return true;
        }
        return false;
    }
};

File: python3/567.py
class Solution:
    def checkInclusion(self, x: str, y: str) -> bool:
        l = len(x)
        m = len(y)
        if l > m:
            return False
        i = [0] * 26
        j = [0] * 26

        def permutation():
            for x, y in zip(i, j):
                if x != y:
                    return False
            return True

        for k in range(l):
            i[ord(x[k]) - 97] += 1
            j[ord(y[k]) - 97] += 1
        if permutation():
            return True
        for k in range(l, m):
            j[ord(y[k - l]) - 97] -= 1
            j[ord(y[k]) - 97] += 1
            if permutation():
                return True
        return False
